### PR TITLE
Core Data: specify context=edit only when needed

### PR DESCRIPTION
**Pull request by creativecoder**
## Description

API requests for fetching entities currently add `context=edit` to all requests. Some endpoints do not use this parameter, which can cause unexpected problems, like the need to preload both the `/` and `/?context=edit` endpoints (which both have exactly the same response).

This change allows specifying query parameters specific to each entity, and adds `context=edit` only to the entities that use that parameter.

This resolves the (rejected) Core ticket (https://core.trac.wordpress.org/ticket/50606) and removes the associated filter for preloading `/?context=edit`.

Follows up on https://github.com/WordPress/gutenberg/pull/22952.

## How has this been tested?

1. Notice that the site icon will load in the editor without prefetching the `/?context=edit` endpoint.
1. Existing resolver unit tests pass with minimal changes.
1. Verify that the `context=edit` parameter is added [only to the endpoints](https://developer.wordpress.org/rest-api/reference/) where it's a valid parameter
1. Verify that the [entity urls generated by the resolver](https://github.com/WordPress/gutenberg/pull/30482/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR113) match the [editor preload paths](https://github.com/WordPress/WordPress/blob/cc0ce2c91d9b59c131012f1d8d648705486bc42b/wp-admin/edit-form-blocks.php#L46).
1. Verify that editor loads and works as expected (no failed API calls or missing data that should be preloaded)

You can also compare [resolver paths for entities](https://github.com/WordPress/gutenberg/pull/30482/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR113) with and without this branch applied, for example:

| Kind | Name | Key | Path (trunk) | Path (this branch) |
| - | - | - | - | - |
| root | __unstableBase | '' | `/?context=edit` | `/` |
| root | site | '' | `/wp/v2/settings/?context=edit` | `/wp/v2/settings` |
| root | postType | post | `/wp/v2/types/post?context=edit` | `/wp/v2/types/post?context=edit` |
| postType | wp_template | tt1-blocks//single | `/wp/v2/templates/tt1-blocks//single?context=edit` | `/wp/v2/templates/tt1-blocks//single?context=edit` |
| postType | wp_global_styles | 26 | `/wp/v2/__experimental/global-styles/26?context=edit` | `/wp/v2/__experimental/global-styles/26?context=edit` |
| root | media | 10 | `/wp/v2/media/10?context=edit` | `/wp/v2/media/10?context=edit` |
| root | taxonomy | category | `/wp/v2/taxonomies/category?context=edit` | `/wp/v2/taxonomies/category?context=edit` |
| root | taxonomy | post_tag | `/wp/v2/taxonomies/post_tag?context=edit` | `/wp/v2/taxonomies/post_tag?context=edit` |
| root | user | 1 | `/wp/v2/users/1?context=edit` | `/wp/v2/users/1?context=edit` |

Note how `/` and `/settings` correctly no longer use `?context=edit`, but other endpoints still do.

## Screenshots 

Site icon displayed in upper left corner:

image

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/core-data/src/entities.js [hunk 1]
```diff
@@ -39,13 +39,15 @@ export const defaultEntities = [
 		kind: 'root',
 		key: 'slug',
 		baseURL: '/wp/v2/types',
+		params: { context: 'edit' },
```
Comment (youknowriad) [inline packages/core-data/src/entities.js]: Should this be  something more like "defaultParams" or something like that? (Don't take my suggestion as a recommendation, I'm no good at naming things, just trying to raise a discussion point)

**Review by youknowriad — COMMENTED** [review]
I think this makes sense.

File: packages/core-data/src/entities.js (hunk 1 above)
Comment (creativecoder) [reply packages/core-data/src/entities.js]: Good question! I updated to `URLparams` and moved them all immediately below `baseURL` to try and make it clearer these are for constructing the endpoint URL.

**Comment by youknowriad** [comment]
Looks like there's a valid test failure. The "types" endpoint does support the context argument.

File: packages/core-data/src/entities.js [hunk 2]
```diff
@@ -39,11 +39,13 @@ export const defaultEntities = [
 		kind: 'root',
 		key: 'slug',
 		baseURL: '/wp/v2/types',
+		URLparams: { context: 'edit' },
```
Comment (Copons) [inline packages/core-data/src/entities.js]: Nitpick, but I'm wondering about the case of this property. 🤔 
It seems we don't have a strong preference for `URL` (I could find all upper, all lower, or capital), but I think we consistently uppercase the first letter after it (e.g. `URLParams`; see also [`URLInput`](https://github.com/WordPress/gutenberg/blob/899286307be533a5643dd1dc168f32d92a204b64/packages/block-editor/src/components/url-input/index.js#L30), or [`__unstableEmbedURLOnPaste`](https://github.com/WordPress/gutenberg/blob/899286307be533a5643dd1dc168f32d92a204b64/packages/block-library/src/paragraph/edit.js#L138), etc.)

**Review by Copons — COMMENTED** [review]
It all seems to work fine, but I'm a bit unclear how I'm supposed to test this. 🤔 
I've tried to compare this branch to trunk, and everything looks exactly the same — with the exception of the `URLparams` property in the `core` state (added to the appropriate elements of `root.entities.config`).

**Comment by creativecoder** [comment]
> Nitpick, but I'm wondering about the case of this property. 

Great point, @Copons. I made another update to the name of this: `baseURLParams`... makes it clear this goes with `baseURL`, and with a capital P for camel casing.

> Looks like there's a valid test failure. The "types" endpoint does support the context argument.

@youknowriad I forgot to update the [test entity object](https://github.com/WordPress/gutenberg/pull/30482/files#diff-a16d857a9f69cfe8a71b293fca645c57a36d4318ab8d1cbe58997a582f9cf221R42) when changing the name of `params`. Should be all set now.

> It all seems to work fine, but I'm a bit unclear how I'm supposed to test this. 🤔

@Copons One thing you can test is if the site icon (when set) shows in the upper left corner of the editor. With the [filtering of `block_editor_preload_paths`](https://github.com/WordPress/gutenberg/pull/30482/files#diff-83705774c15f7e3df5b54b2096fe476721e8b962ba319a8e7dc46d7b9b0b27b4L159) removed in this PR, fetching the site icon url would otherwise fail.

Similarly, loading the editor fails (blank white screen) when the entity api paths don't match the critical [preload paths](https://github.com/WordPress/WordPress/blob/cc0ce2c91d9b59c131012f1d8d648705486bc42b/wp-admin/edit-form-blocks.php#L46). I know testing for "not failure" isn't ideal, but one thing you can do is verify that the [entity urls generated by the resolver](https://github.com/WordPress/gutenberg/pull/30482/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR113) match those preload paths (just by console.log, if nothing else!)

image

**Comment by TimothyBJacobs** [comment]
This approach looks great to me. I haven't tested for regressions though.

**Review by Copons — APPROVED** [review]
Thanks for the pointers @creativecoder!

I already tested in roughly the same ways you suggested and it worked as expected; I also haven't noticed any regressions in either post and site editors, so it should be good to go! 🙂

**Comment by creativecoder** [comment]
Note that I've added a new unit test for the `getEntityRecord` resolver when a `query` is provided, and done some additional confidence checks by comparing the entity api paths used on this branch with `trunk` in the post and site editors (see PR description for details).

Any other suggestions for regression testing this? Otherwise I think it's ready to merge.

**Comment by gziolo** [comment]
Does it require any changes in WordPress core? The list of preloaded paths is here:

https://github.com/WordPress/wordpress-develop/blob/e4bda140d4437f5bd5690e18bc12a0eb70bb92c9/src/wp-admin/edit-form-blocks.php#L48-L59

**Comment by creativecoder** [comment]
> Does it require any changes in WordPress core?

Good question! I don't think so: those existing preload paths match what core-data is using. A main goal of this PR was to remove the `?context=edit` from the `/` path, to match the existing preload path in Core.

**Comment by gziolo** [comment]
@creativecoder, nice. Thank you for checking 👍🏻

**Pull request merged**